draft-ietf-quic-transport.md

```diff
@@ -5874,6 +5874,288 @@ decisions are made independently of client-selected values; a Source Connection
+
+An on-path attacker can also replace the addresses of packets on either side and
+therefore cause the client or server to have an incorrect view of the remote
+addresses.
+
+#### Parameter Negotiation
+
+The entire handshake is cryptographically protected, with the Initial packets
+being encrypted with per-version keys and the Handshake and later packets being
+encrypted with keys derived from the TLS key exchange.  Further, parameter
+negotiation is folded into the TLS transcript and thus provides the same
+security guarantees as ordinary TLS negotiation.  Thus, an attacker can observe
+the client's transport parameters (as long as it knows the QUIC version-specific
+salt) but cannot observe the server's transport parameters and cannot influence
+parameter negotiation.
+
+Connection IDs are unencrypted but integrity protected in all messages.  They
+are not incorporated in the TLS handshake transcript.
+
+This version of QUIC does not incorporate a version negotiation mechanism;
+implementations of QUIC with incompatible versions will simply fail to
+negotiate.
+
+
+### Protected Packets {#protected-packet-properties}
+
+Packet protection ({{packet-protected}}) provides authentication and encryption
+of all QUIC packets except Version Negotiation and Retry packets, see
+{{QUIC-TLS}} for more details.  This section considers passive and active
+attacks against protected packets.
+
+Both on-path and off-path attackers can mount a passive attack in which they
+save observed QUIC packets for an offline attack against QUIC packet protection
+at a future time; this is true for any observer of any packet on any network.
+
+A blind attacker, one who injects packets without being able to observe valid
+packets for a QUIC connection, is unlikely to be successful, since packet
+protection ensures that valid packets are only generated by endpoints which
+possess the key material established during the handshake; see {{handshake}} and
+{{handshake-properties}}.  Similarly, any active attacker that observes QUIC
+packets and attempts to insert new data or modify existing data in those packets
+should not be able to generate packets deemed valid by the receiving endpoint.
+
+A spoofing attack, in which an active attacker rewrites unprotected parts of a
+QUIC packet that it forwards or injects, such as the source or destination
+address, is only effective if the attacker can forward packets to the original
+endpoint.  Packet protection ensures that the packet payloads can only be
+processed by the endpoints that completed the handshake, and invalid QUIC
+packets are ignored by those endpoints.
+
+An attacker can also modify the boundaries between QUIC packets and UDP
+datagrams, causing multiple packets to be coalesced into a single datagram, or
+splitting coalesced packets into multiple datagrams.  Such modification has no
+functional effect on a QUIC connection, although it might change the performance
+characteristics exhibited by the receiving endpoint.
```

Comment:
Good observation.  I remember having discussions about heuristics that assume packets sent in a single datagram arrived together.  This means those heuristics aren't reliable.  I don't recall whether we endorsed such approaches in the document (PMTU probing, maybe?), but this suggests that we might want to explicitly caution against them.  (Not necessarily here.)

Reply:
Of note here is that we allow Initial packets to be smaller than 1200 bytes, as long as they are coalesced with other stuff to make up a 1200 byte UDP datagram.  That too is vulnerable to this sort of composition problem.  I think that it's fine to make these assumptions and risk modification as long as they are recognized as such.  For Initial, we are doing so on the understanding that elements on the path are able to disrupt the handshake at that point anyway.  For PMTUD, we might make a different assessment as the conditions are different.

Reply:
Good thinking, it might be worth a note around those rules in the Initial that such a thing is possible (and might be more likely than if you just padded the initial packet?). PMTUD is probably worth calling out a little bit more clearly -- I'll file an editorial issue for each, since that's way elsewhere.